### DISCUSSION TITLE
Speeding up the creation of an Entity Manager

### DISCUSSION THREAD

**Discussion by rubenporras** [eclipse-ee4j/eclipselink#2112] [Q&A]
Hi,

we have problems with multiple threads creating entity managers the first time a persistence unit is used, the overall picture is that many threads are waiting for more than one minute on EntityManagerFactoryDelegate.getAbstractSession (we have 64 threads running concurrently), an example of one of such 63 threads:

```
"request-id-1194" Id=312 BLOCKED on org.eclipse.persistence.internal.jpa.EntityManagerFactoryDelegate@2102c7ca owned by "request-id-1188" Id=306
	at app//org.eclipse.persistence.internal.jpa.EntityManagerFactoryDelegate.getAbstractSession(EntityManagerFactoryDelegate.java:215)
	-  blocked on org.eclipse.persistence.internal.jpa.EntityManagerFactoryDelegate@2102c7ca
	at app//org.eclipse.persistence.internal.jpa.EntityManagerFactoryDelegate.createEntityManagerImpl(EntityManagerFactoryDelegate.java:330)
	at app//org.eclipse.persistence.internal.jpa.EntityManagerFactoryImpl.createEntityManagerImpl(EntityManagerFactoryImpl.java:350)
	at app//org.eclipse.persistence.internal.jpa.EntityManagerFactoryImpl.createEntityManager(EntityManagerFactoryImpl.java:313)
	at app//com.<xxx>.EntityManagerCreator.createEntityManager(EntityManagerCreator.java:57)
```

And then the thread which is blocking:

> "request-id-1188" Id=306 RUNNABLE
> 	at app//org.eclipse.persistence.internal.sessions.AbstractSession.getLogSessionString(AbstractSession.java:2717)
> 	at app//org.eclipse.persistence.logging.AbstractSessionLog.getSessionString(AbstractSessionLog.java:821)
> 	at app//org.eclipse.persistence.logging.AbstractSessionLog.getSupplementDetailString(AbstractSessionLog.java:791)
> 	at app//org.eclipse.persistence.logging.DefaultSessionLog.log(DefaultSessionLog.java:141)
> 	at app//org.eclipse.persistence.internal.sessions.AbstractSession.log(AbstractSession.java:3659)
> 	at app//org.eclipse.persistence.internal.sessions.AbstractSession.log(AbstractSession.java:4902)
> 	at app//org.eclipse.persistence.internal.sessions.AbstractSession.log(AbstractSession.java:4874)
> 	at app//org.eclipse.persistence.internal.sessions.AbstractSession.log(AbstractSession.java:4850)
> 	at app//org.eclipse.persistence.internal.sessions.AbstractSession.log(AbstractSession.java:4772)
> 	at app//org.eclipse.persistence.internal.sessions.DatabaseSessionImpl.preConnectDatasource(DatabaseSessionImpl.java:853)
> 	at app//org.eclipse.persistence.internal.sessions.DatabaseSessionImpl.loginAndDetectDatasource(DatabaseSessionImpl.java:809)
> 	at app//org.eclipse.persistence.internal.jpa.EntityManagerFactoryProvider.login(EntityManagerFactoryProvider.java:261)
> 	at app//org.eclipse.persistence.internal.jpa.EntityManagerSetupImpl.deploy(EntityManagerSetupImpl.java:771)
> 	at app//org.eclipse.persistence.internal.jpa.EntityManagerFactoryDelegate.getAbstractSession(EntityManagerFactoryDelegate.java:222)
> 	at app//org.eclipse.persistence.internal.jpa.EntityManagerFactoryDelegate.createEntityManagerImpl(EntityManagerFactoryDelegate.java:330)
> 	at app//org.eclipse.persistence.internal.jpa.EntityManagerFactoryImpl.createEntityManagerImpl(EntityManagerFactoryImpl.java:350)
> 	at app//org.eclipse.persistence.internal.jpa.EntityManagerFactoryImpl.createEntityManager(EntityManagerFactoryImpl.java:313)
> 	at app//com.<xxx>.EntityManagerCreator.createEntityManager(EntityManagerCreator.java:57)

My first question, is such slowness by the creating of an EntityManager expected under some configuration? To my knowledge the database is performing just fine.

I have been looking at how one could improve that, and by looking at the codebase, I would have tried to set in the persistence unit the following properties:

```
      <property name="eclipselink.target-database" value="org.eclipse.persistence.platform.database.oracle.Oracle19Platform"/>
      <property name="eclipselink.validation-only" value="true"/>

```

 Does it make sense to do so if we are sure we are running on an Oracle 19 database? Are there any drawbacks other that we disable the automatic configuration? I could not find any documentation that list any drawbacks.

Kind Regards

**Comment by chris-delahunt** [comment]
"eclipselink.validation-only" was created for validation only purposes, not for a usable EntityManagerFactory/EntityManager creation. It was added I believe to validate the mappings, but a usable persistence context needs the connections initialized. 

"eclipselink.target-database" is certainly helpful as the auto config will use the driver metadata to determine the connected database and then the platform used. I can't say how much overhead auto db platform lookups adds, but it is unnecessary overhead if you know the platform you want it to be using anyway - and you can override it in properties for testing should you wish to run against an in-memory or other database.

Otherwise, I'd say the best option is to try to setup the EMF and obtain an EntityManager after application initialization or deployment, before the app tries obtaining them. See https://eclipse.dev/eclipselink/documentation/2.5/jpa/extensions/p_deploy_on_startup.htm for details.
